prefect_kubernetes/jobs.py

```diff
@@ -299,3 +310,217 @@ def kubernetes_orchestrator():
+                    core_v1_client.list_namespaced_pod,
+                    namespace=self._kubernetes_job.namespace,
+                    label_selector=pod_selector,
+                    **self._kubernetes_job.api_kwargs,
+                )
+                for pod in v1_pod_list.items:
+                    pod_name = pod.metadata.name
+
+                    if (
+                        pod.status.phase == "Pending"
+                        or pod_name in self.pod_logs.keys()
+                    ):
+                        continue
+
+                    self.logger.info(f"Capturing logs for pod {pod_name!r}.")
+
+                    self.pod_logs[pod_name] = await run_sync_in_worker_thread(
+                        core_v1_client.read_namespaced_pod_log,
+                        namespace=self._kubernetes_job.namespace,
+                        name=pod_name,
+                        container=latest_v1_job.spec.template.spec.containers[0].name,
+                        **self._kubernetes_job.api_kwargs,
+                    )
+
+                if latest_v1_job.status.active:
+                    await sleep(self._kubernetes_job.interval_seconds)
+                    if self._kubernetes_job.timeout_seconds:
+                        elapsed_time += self._kubernetes_job.interval_seconds
+                elif latest_v1_job.status.failed:
+                    raise RuntimeError(
+                        f"Job {latest_v1_job.metadata.name!r} failed, check the "
+                        "Kubernetes pod logs for more information."
+                    )
+                elif latest_v1_job.status.succeeded:
+                    completed = True
+                    self.logger.info(
+                        f"Job {latest_v1_job.metadata.name!r} has completed."
+                    )
+
+    async def fetch_result(self) -> Dict[str, Any]:
+        """Fetch the results of the job.
+
+        Returns:
+            The logs from each of the pods in the job.
+        """
+        if self._kubernetes_job.delete_after_completion:
+            with self._kubernetes_job.credentials.get_client(
+                "batch"
+            ) as batch_v1_client:
+                deleted_v1_job = await run_sync_in_worker_thread(
+                    batch_v1_client.delete_namespaced_job,
+                    namespace=self._kubernetes_job.namespace,
+                    name=self._kubernetes_job.v1_job.metadata.name,
+                    **self._kubernetes_job.api_kwargs,
+                )
+                self.logger.info(
+                    f"Job {self._kubernetes_job.v1_job.metadata.name} deleted "
+                    f"with {deleted_v1_job.status!r}."
+                )
+
+        if self.pod_logs is None:
+            raise ValueError(
+                "`KubernetesJobRun.wait_for_completion` was never called - "
+                "therefore no pod logs were collected and they cannot be returned."
+            )
```

Comment:
I think that we could track the state of the job in a private attribute so that this logic is more explicit than implicit. Then we can give a message along the lines of, "This job is in non terminal state {state}, unable to retrieve results. Try again later."